File: 001_MergeStringsAlternately.cpp
// Topic: Two Pointers / String
// Merge Strings Alternately: Combine two strings by adding letters in alternating order.
// Time Complexity: O(n) where n = s1.size() + s2.size()
// Space Complexity: O(n) for the resulting string

#include <bits/stdc++.h>
using namespace std;

string mergeAlternately(const string& s1, const string& s2)
{
    string res;
    res.reserve(s1.size() + s2.size());
    size_t i = 0, j = 0;
    while (i < s1.size() || j < s2.size())
    {
        if (i < s1.size())
            res.push_back(s1[i++]);
        if (j < s2.size())
            res.push_back(s2[j++]);
    }
    return res;
}

int main()
{
    cout << mergeAlternately("abc", "pqr") << "\n";  // apbqcr
    cout << mergeAlternately("ab", "pqrs") << "\n";  // apbqrs
    cout << mergeAlternately("abcd", "pq") << "\n";  // apbqcd
    return 0;
}

File: 001_MergeStringsAlternately.py
"""
Topic: Two Pointers / String
Merge Strings Alternately: merge characters from two strings alternately
Time Complexity: O(n + m)
Space Complexity: O(n + m)
"""

def merge_alternately(a: str, b: str) -> str:
    i = j = 0
    res = []
    turn = 0
    while i < len(a) or j < len(b):
        if turn == 0 and i < len(a):
            res.append(a[i]); i += 1
        elif turn == 1 and j < len(b):
            res.append(b[j]); j += 1
        else:
            # if current string exhausted, append from the other
            if i < len(a): res.append(a[i]); i += 1
            elif j < len(b): res.append(b[j]); j += 1
        turn ^= 1
    return ''.join(res)


if __name__ == '__main__':
    print(merge_alternately('abc', 'pqr'))  # apbqcr
    print(merge_alternately('ab', 'pqrs'))  # apbqrs
    print(merge_alternately('', 'xyz'))     # xyz
